src/StoragelessSession/Http/SessionMiddleware.php

```diff
@@ -217,18 +218,32 @@ public function extractSessionContainer(Token $token = null) : Data
     /**
      * @param Data     $sessionContainer
      * @param Response $response
+     * @param Token    $originalToken
      *
      * @return Response
      *
      * @throws \InvalidArgumentException
      */
-    private function appendToken(Data $sessionContainer, Response $response) : Response
+    private function appendToken(Data $sessionContainer, Response $response, Token $originalToken = null) : Response
     {
-        if ($sessionContainer->isEmpty()) {
+        $empty = $sessionContainer->isEmpty();
+
+        if ($empty && null === $originalToken) {
```

Comment:
I think that relying on the parsed, validated and verified token is better because we would be certain that we're not using garbage (that could be stored in the cookie and mess things up).